xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -1315,6 +1374,172 @@ private void cleanUpResourceTimers() {
+      String resourceName = entry.getKey();
+      ResourceSubscriber subscriber = entry.getValue();
+
+      // Attach error details to the subscribed resources that included in the ADS update.
+      if (unpackedResourceNames.contains(resourceName)) {
+        subscriber.onRejected(version, updateTime, errorDetail);
+      }
+    }
+  }
+
+  private static String combineErrors(List<String> errors) {
+    return Joiner.on('\n').join(errors);
+  }
+
+  /**
+   * Captures ResourceSubscriber metadata, used by the xDS config dump.
+   */
+  static final class ResourceMetadata {
+    private final String version;
+    private final ResourceMetadataStatus status;
+    private final long updateTime;
+    @Nullable private final Any rawResource;
+    @Nullable private final UpdateFailureState errorState;
+
+    private ResourceMetadata(
+        ResourceMetadataStatus status, String version, long updateTime, @Nullable Any rawResource,
+        @Nullable UpdateFailureState errorState) {
+      this.status = checkNotNull(status, "status");
+      this.version = checkNotNull(version, "version");
+      this.updateTime = updateTime;
+      this.rawResource = rawResource;
+      this.errorState = errorState;
+    }
+
+    static ResourceMetadata newResourceMetadataUnknown() {
+      return new ResourceMetadata(ResourceMetadataStatus.UNKNOWN, "", 0, null, null);
+    }
+
+    static ResourceMetadata newResourceMetadataRequested() {
+      return new ResourceMetadata(ResourceMetadataStatus.REQUESTED, "", 0, null, null);
+    }
+
+    static ResourceMetadata newResourceMetadataDoesNotExist() {
+      return new ResourceMetadata(ResourceMetadataStatus.DOES_NOT_EXIST, "", 0, null, null);
+    }
+
+    static ResourceMetadata newResourceMetadataAcked(
+        Any resource, String version, long updateTime) {
+      checkNotNull(resource, "resource");
+      return new ResourceMetadata(
+          ResourceMetadataStatus.ACKED, version, updateTime, resource, null);
+    }
+
+    static ResourceMetadata newResourceMetadataNacked(
+        ResourceMetadata metadata, String failedVersion, long failedUpdateTime,
+        String failedDetails) {
+      checkNotNull(metadata, "metadata");
+      return new ResourceMetadata(ResourceMetadataStatus.NACKED,
+          metadata.getVersion(), metadata.getUpdateTime(), metadata.getRawResource(),
+          new UpdateFailureState(failedVersion, failedUpdateTime, failedDetails));
+    }
+
+    String getVersion() {
+      return version;
+    }
+
+    ResourceMetadataStatus getStatus() {
+      return status;
+    }
+
+    long getUpdateTime() {
+      return updateTime;
+    }
+
+    @Nullable
+    Any getRawResource() {
+      return rawResource;
+    }
+
+    @Nullable
+    UpdateFailureState getErrorState() {
+      return errorState;
+    }
+
+    enum ResourceMetadataStatus {
+      UNKNOWN, REQUESTED, DOES_NOT_EXIST, ACKED, NACKED
```

Comment:
nit: Shall we ensure the enums here match the proto-defined enums? E.g., through assertions.

Reply:
I like this idea! At this moment there's no CSDS protos/dependencies imported though. I'll add this test to the next PR with CSDS service.